### PR TITLE
Feature/msgpacktoken

### PR DESCRIPTION
**Pull request by roman-kozachenko**
https://gist.github.com/roman-kozachenko/d0fff9c0d1e1d9e0d1c17211ff393ab0

### CODE REVIEW

File: src/msgpack.light/BaseMsgPackReader.cs [hunk 1]
```diff
@@ -66,129 +74,16 @@ public virtual DataTypes ReadDataType()
...
-
-        private static uint? TryGetLengthFromFixStr(DataTypes type)
-        {
-            var length = type - DataTypes.FixStr;
-            return type.GetHighBits(3) == DataTypes.FixStr.GetHighBits(3) ? length : (uint?)null;
+            MsgPackTokenConverter.ReadToken(this, _skippingTokenBuilder);
         }
 
-        protected static uint? TryGetLengthFromFixArray(DataTypes type)
+        internal static uint? TryGetLengthFromFixArray(DataTypes type)
```
Comment (aensidhe) [inline src/msgpack.light/BaseMsgPackReader.cs]: `protected internal` ?

File: src/msgpack.light/MsgPackByteArrayReader.cs [hunk 2]
```diff
@@ -9,7 +9,7 @@ internal class MsgPackByteArrayReader : BaseMsgPackReader
 
         private uint _offset;
 
-        public MsgPackByteArrayReader(byte[] data)
+        public MsgPackByteArrayReader(byte[] data) : base()
```
Comment (aensidhe) [inline src/msgpack.light/MsgPackByteArrayReader.cs]: Unnecessary change.

File: src/msgpack.light/MsgPackToken.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿namespace ProGaudi.MsgPack.Light
+{
+    public class MsgPackToken
+    {
+        public DataTypes DataType { get; }
+
+        public byte[] TokenBytes { get; }
```
Comment (aensidhe) [inline src/msgpack.light/MsgPackToken.cs]: `Bytes` or `Token` or `Value`.

```
MsgPackToken token;
token.TokenBytes // <--- strange
```

File: src/msgpack.light/TokenBuilders/IMsgPackTokenBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿namespace ProGaudi.MsgPack.Light.TokenBuilders
+{
+    internal interface IMsgPackTokenBuilder
+    {
+        void Add(DataTypes dataType);
+
+        MsgPackToken CreateToken(DataTypes dataType);
```
Comment (aensidhe) [inline src/msgpack.light/TokenBuilders/IMsgPackTokenBuilder.cs]: Do we really need this construct? If someone has a type token and byte array, he will always able to deserialize it.

Comment (roman-kozachenko) [reply src/msgpack.light/TokenBuilders/IMsgPackTokenBuilder.cs]: As we discussed, MsgPackTokenConverter.Read method is very similar to BaseMsgPackReader.Skip method.
The only difference is that In Read method we collect bytes to buffer and made some operations with it (.ToArray() for example),  but in Skip method we don't do that.
To avoid logic duplication, IMsgPackTokenBuilder introduced.

File: src/msgpack.light/MsgPackToken.cs (hunk 3 above)
Comment (roman-kozachenko) [reply src/msgpack.light/MsgPackToken.cs]: Renamed to RawValue
